lib/logstash/outputs/newrelic.rb

```diff
@@ -101,19 +112,29 @@ def multi_receive(events)
           }
         }
       },
-      :logs => payload
+      :logs => nr_logs
     }
-    @semaphor.acquire()
+
     execute = @executor.java_method :submit, [java.lang.Runnable]
     execute.call do
-      begin
-        io = StringIO.new
-        gzip = Zlib::GzipWriter.new(io)
-        gzip << [payload].to_json
-        gzip.close
-        nr_send(io.string)
-      ensure
-        @semaphor.release()
+      compressed_payload = StringIO.new
+      gzip = Zlib::GzipWriter.new(compressed_payload)
+      gzip << [payload].to_json
+      gzip.close
+
+      compressed_size = compressed_payload.string.bytesize
+      log_record_count = nr_logs.length
+
+      if compressed_size >= MAX_PAYLOAD_SIZE_BYTES && log_record_count == 1
+        @logger.error("Can't compress record below required maximum packet size and it will be discarded.")
+      elsif compressed_size >= MAX_PAYLOAD_SIZE_BYTES && log_record_count > 1
+        @logger.debug("Compressed payload size (#{compressed_size}) exceededs maximum packet size (1MB) and will be split in two.")
+        split_index = log_record_count / 2
+        package_and_send_recursively(nr_logs[0...split_index])
+        package_and_send_recursively(nr_logs[split_index..-1])
```

Comment:
I'd like to have a test included somehow that checks that our split logic here is precisely right and doesn't have any off-by-ones or anything. Is there a way to do that?

Reply:
Good point, let me work on that. I can regenerate the log test files and include a msgId (1, 2, 3...) in each of them, and then have our test verify that all the msgIds were received.

Reply:
Done! I've modified the test so it verifies that all messages were received. I also ensured to send an uneven amount of logs, to force them being divided into unequally long arrays.